_drafts/2021-05-04-opaque-pointers.md

```diff
@@ -0,0 +1,308 @@
+description: "An overview of the opaque pointer pattern in C. How to create objects in C."
+author: nickmiller
+---
+
+I've written a lot of C++ in my career, but I still prefer to design in
+C for most embedded projects ("why" is the subject of a much longer rant-filled post).
+I'm not a big proponent of OOP in general,
+but I do think having an "instance" of something which contains stateful data is a generally
+useful thing for embedded software. For example, a ring buffer instance (aka circular FIFO queue),
+of which there might be many in the system. Each instance contains stateful data,
+like the current position of read and write pointers. What's the best way
+to model this in C?
+
+<!-- excerpt start -->
+
+Objects are not a native concept in C, but you can achieve something
+resembling objects by using a design pattern known as the "opaque pointer".
+This post will show you what the pattern is, explain some of the finer
+details, and provide guidance on when you should (and shouldn't) use it.
+
+<!-- excerpt end -->
+
+The opaque pointer is one of the most useful and frequently-encountered design
+patterns in C.
+Even if you are a die-hard C++ proponent (or *gasp* Rust!), the reality is
+that there are, and will continue to be, a lot of embedded projects out there
+written in C. And you may need to contribute to one of them.
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## The Opaque Pointer Pattern
+
+*opaque: not able to be seen through; not transparent.*
+
+In traditional OOP, an object has state (data) and behavior (functions) coupled together in
+one class.
+
+If you want to have something resembling objects in C, you can:
+
+1. Define a struct with your state data
+2. Define functions which take a pointer to the struct as the first parameter
+3. Declare a variable with that struct type to create an instance of the "object"
+
+Need another object instance? Declare another variable with that struct type.
+Boom! Objects in C, baby!
+
+Of course, there's a little more to it than that (but not much).
+
+Before getting into the details, let me show you the complete pattern first, using a ring buffer
+that stores `uint8_t` as an example.
+This is not a complete ringbuffer implementation.
+It's just enough to show you the general pattern.
+
+`ringbuffer.h`
+
+```c
+#pragma once
+
+#include <stdint.h>
+#include <stdbool.h>
+
+// Opaque pointer type to represent a ringbuffer instance
+typedef struct ringbuffer_instance_t* ringbuffer_t;
+
+// Functions that operate on an instance
+ringbuffer_t ringbuffer_create(uint32_t capacity);
+uint32_t ringbuffer_capacity(ringbuffer_t instance);
+bool ringbuffer_enqueue(ringbuffer_t instance, uint8_t item);
+bool ringbuffer_dequeue(ringbuffer_t instance, uint8_t* item);
+void ringbuffer_destroy(ringbuffer_t instance);
+```
+
+`ringbuffer.c`
+
+```c
+#include "ringbuffer.h"
+
+// Private struct, only accessible from within this file
```

Comment:
People ask me often where they should put struct definitions that need to be shared across multiple files but within the same "module". I generally say to have a header file, `ringbuffer_private.h` and keep it in there. 

If you have your own thoughts on this, might be useful to add a sentence or two.